BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Authentication/ReflectionCredentialEnvelope.cs

```diff
@@ -0,0 +1,244 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility.Implementation.Authentication
+{
+    using System;
+    using System.Linq.Expressions;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    /// <summary>
+    /// This is an envelope for an instance of Azure.Core.TokenCredential.
+    /// This class uses reflection to interact with the Azure.Core library.
+    /// </summary>
+    /// <remarks>
+    /// Our SDK currently targets net452, net46, and netstandard2.0.
+    /// Azure.Core.TokenCredential is only available for netstandard2.0.
+    /// </remarks>
+    internal class ReflectionCredentialEnvelope : ICredentialEnvelope
+    {
+        private readonly object tokenCredential;
+        private readonly object tokenRequestContext;
+
+        /// <summary>
+        /// Create an instance of <see cref="ReflectionCredentialEnvelope"/>.
+        /// </summary>
+        /// <param name="tokenCredential">An instance of Azure.Core.TokenCredential.</param>
+        public ReflectionCredentialEnvelope(object tokenCredential)
+        {
+            this.tokenCredential = tokenCredential ?? throw new ArgumentNullException(nameof(tokenCredential));
+
+            if (tokenCredential.GetType().IsSubclassOf(Type.GetType("Azure.Core.TokenCredential, Azure.Core")))
+            {
+                this.tokenRequestContext = AzureCore.MakeTokenRequestContext(scopes: CredentialConstants.GetScopes());
+            }
+            else
+            {
+                throw new ArgumentException($"The provided {nameof(tokenCredential)} must inherit Azure.Core.TokenCredential", nameof(tokenCredential));
+            }
+        }
+
+        /// <summary>
+        /// Gets the TokenCredential instance held by this class.
+        /// </summary>
+        public object Credential => this.tokenCredential;
+
+        /// <summary>
+        /// Gets an Azure.Core.AccessToken.
+        /// </summary>
+        /// <param name="cancellationToken">The System.Threading.CancellationToken to use.</param>
+        /// <returns>A valid Azure.Core.AccessToken.</returns>
+        public string GetToken(CancellationToken cancellationToken = default)
+        {
+            SdkInternalOperationsMonitor.Enter();
```

Comment:
it may be better to leave the job of entering/exiting internal operation to the consumer (like Transmission class)

Reply:
Will investigate this in second PR.